client/tso_request_dispatcher.go

```diff
@@ -0,0 +1,751 @@
+			continue
+		}
+		log.Error("[pd] create the tso stream failed", zap.String("dc", dc), zap.String("addr", addr), errs.ZapError(err))
+		cancel()
+	}
+	return nil
+}
+
+func extractSpanReference(tbc *tsoBatchController, opts []opentracing.StartSpanOption) []opentracing.StartSpanOption {
+	for _, req := range tbc.getCollectedRequests() {
+		if span := opentracing.SpanFromContext(req.requestCtx); span != nil {
+			opts = append(opts, opentracing.ChildOf(span.Context()))
+		}
+	}
+	return opts
+}
+
+func (c *client) processTSORequests(stream tsoStream, dcLocation string, tbc *tsoBatchController, opts []opentracing.StartSpanOption) error {
+	if len(opts) > 0 {
+		span := opentracing.StartSpan("pdclient.processTSORequests", opts...)
+		defer span.Finish()
+	}
+
+	requests := tbc.getCollectedRequests()
+	count := int64(len(requests))
+	physical, logical, suffixBits, err := stream.processRequests(c.GetClusterID(c.ctx), dcLocation, requests, tbc.batchStartTime)
+	if err != nil {
+		c.finishTSORequest(requests, 0, 0, 0, err)
+		return err
+	}
+	// `logical` is the largest ts's logical part here, we need to do the subtracting before we finish each TSO request.
+	firstLogical := addLogical(logical, -count+1, suffixBits)
+	c.compareAndSwapTS(dcLocation, physical, firstLogical, suffixBits, count)
+	c.finishTSORequest(requests, physical, firstLogical, suffixBits, nil)
+	return nil
+}
+
+// Because of the suffix, we need to shift the count before we add it to the logical part.
+func addLogical(logical, count int64, suffixBits uint32) int64 {
+	return logical + count<<suffixBits
+}
+
+func (c *client) compareAndSwapTS(dcLocation string, physical, firstLogical int64, suffixBits uint32, count int64) {
+	largestLogical := addLogical(firstLogical, count-1, suffixBits)
+	lastTSOInterface, loaded := c.lastTSMap.LoadOrStore(dcLocation, &lastTSO{
+		physical: physical,
+		// Save the largest logical part here
+		logical: largestLogical,
+	})
+	if !loaded {
+		return
+	}
+	lastTSOPointer := lastTSOInterface.(*lastTSO)
+	lastPhysical := lastTSOPointer.physical
+	lastLogical := lastTSOPointer.logical
+	// The TSO we get is a range like [largestLogical-count+1, largestLogical], so we save the last TSO's largest logical to compare with the new TSO's first logical.
+	// For example, if we have a TSO resp with logical 10, count 5, then all TSOs we get will be [6, 7, 8, 9, 10].
+	if tsLessEqual(physical, firstLogical, lastPhysical, lastLogical) {
+		panic(errors.Errorf("%s timestamp fallback, newly acquired ts (%d, %d) is less or equal to last one (%d, %d)",
+			dcLocation, physical, firstLogical, lastPhysical, lastLogical))
+	}
+	lastTSOPointer.physical = physical
+	// Same as above, we save the largest logical part here.
+	lastTSOPointer.logical = largestLogical
+}
+
+func tsLessEqual(physical, logical, thatPhysical, thatLogical int64) bool {
+	if physical == thatPhysical {
+		return logical <= thatLogical
+	}
+	return physical < thatPhysical
+}
```

Comment:
if move to tso util, it will be better

Reply:
there is only one place to use these functions after refactor, so it isn't necessary to move it to tso util for now, is it?